File: array/binsearch.go
package main

import (
	"fmt"
)

func binsearch(slice []int, target int) int {
	start := 0
	end := len(slice)
	for start <= end {
		middle := start + (end-start)/2
		if slice[middle] > target {
			end = middle - 1
		} else if slice[middle] < target {
			start = middle + 1
		} else {
			return middle
		}
	}
	return 0
}
func main() {
	slice := []int{11, 11, 22, 33, 44, 55, 66, 77, 88}
	res := binsearch(slice, 22)
	fmt.Printf("%d\n", res)
}

File: algorithm/Binsearch.java
class Binsearch {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length -1;
        while (left <= right){
            int middle = (left+right)/2;
            if(nums[middle] > target){
                right = middle-1;
            }else if(nums[middle] < target){
                left = middle+1;
            }else{
                return middle;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = new int[] { 12, 23, 34, 45, 56, 67, 77, 89, 90 };
        Binsearch s = new Binsearch();
        s.search(arr,23);
    }
}